Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTP1Connections.swift

```diff
@@ -0,0 +1,438 @@
+            guard connectionState.isAvailable else {
+                // connection is not available anymore, we may have just leased it for a request
+                return nil
+            }
+
+            self.connections.remove(at: index)
+            self.overflowIndex = self.connections.index(before: self.overflowIndex)
+            return connectionState
+        }
+
+        mutating func leaseConnection(_ connectionID: Connection.ID) -> Connection {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("A connection that we don't know was released? Something is very wrong...")
+            }
+
+            return self.connections[index].lease()
+        }
+
+        mutating func leaseAvailableConnection(onPreferred preferredEL: EventLoop) -> Connection? {
+            guard let index = self.findAvailableConnection(onPreferred: preferredEL) else {
+                return nil
+            }
+
+            return self.connections[index].lease()
+        }
+
+        mutating func leaseAvailableConnection(onRequired preferredEL: EventLoop) -> Connection? {
+            guard let index = self.findAvailableConnection(onRequired: preferredEL) else {
+                return nil
+            }
+
+            return self.connections[index].lease()
+        }
+
+        @discardableResult
+        mutating func removeConnection(_ connectionID: Connection.ID) -> HTTP1ConnectionState {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("We tried to create a new connection, that we know nothing about?")
+            }
+            let connectionState = self.connections.remove(at: index)
+            self.overflowIndex = self.connections.index(before: self.overflowIndex)
+            return connectionState
+        }
+
+        mutating func shutdown() -> StateMachine.ConnectionAction.CleanupContext {
+            var cleanupContext = StateMachine.ConnectionAction.CleanupContext()
+
+            self.connections = self.connections.compactMap { connectionState in
+                var connectionState = connectionState
+                if connectionState.cleanup(&cleanupContext) {
+                    return nil
+                }
+                return connectionState
+            }
+
+            return cleanupContext
+        }
+
+        private func findAvailableConnection(onPreferred preferredEL: EventLoop) -> Int? {
+            var eventLoopMatch: (Int, NIODeadline)?
+            var goodMatch: (Int, NIODeadline)?
+
+            // To find an appropriate connection we iterate all existing connections.
+            // While we do this we try to find the best fitting connection for our request.
+            //
+            // A perfect match, runs on the same eventLoop and has been idle the shortest amount
+            // of time.
```

Comment:
Might be worth adding that "idle the shortest amount of time" means "returned the most recently".